newapi/ooniapi/citizenlab.py

```diff
@@ -0,0 +1,822 @@
+
+    def push_to_repo(self, username):
+        log.debug("pushing branch to GitHub")
+        self.repo.remotes.rworigin.push(
+            self.get_user_branchname(username),
+            progress=ProgressPrinter(),
+            force=True,
+        )
+
+    def propose_changes(self, username: str) -> str:
+        with self.get_user_lock(username):
+            log.debug("proposing changes")
+            self.set_state(username, "PR_OPEN")
+            self.push_to_repo(username)
+            pr_id = self.open_pr(self.get_user_branchname(username))
+            self.set_pr_id(username, pr_id)
+            return pr_id
+
+
+class BadURL(HTTPException):
+    code = 400
+    description = "Invalid URL"
+
+
+class BadCategoryCode(HTTPException):
+    code = 400
+    description = "Invalid category code"
+
+
+class BadCategoryDescription(HTTPException):
+    code = 400
+    description = "Invalid category description"
+
+
+class BadDate(HTTPException):
+    code = 400
+    description = "Invalid date"
+
+
+def check_url(url):
+    if not VALID_URL.match(url):
+        raise BadURL()
+    elif any([c in url for c in BAD_CHARS]):
+        raise BadURL()
+    elif url != url.strip():
+        raise BadURL()
+    elif urlparse(url).path == "":
+        raise BadURL()
+
+
+def validate_entry(entry: List[str]) -> None:
+    url, category_code, category_desc, date_str, user, notes = entry
+    check_url(url)
+    if category_code not in CATEGORY_CODES:
+        raise BadCategoryCode()
+    if category_desc != CATEGORY_CODES[category_code]:
+        raise BadCategoryDescription()
+    try:
+        d = datetime.strptime(date_str, "%Y-%m-%d").date().isoformat()
+        if d != date_str:
+            raise BadDate()
+    except Exception:
+        raise BadDate()
+
+
+def get_username():
+    return request._user_nickname
+
+
+def get_url_list_manager():
+    conf = current_app.config
+    return URLListManager(
+        working_dir=Path(conf["GITHUB_WORKDIR"]),
+        github_token=conf["GITHUB_TOKEN"],
+        origin_repo=conf["GITHUB_ORIGIN_REPO"],
+        push_repo=conf["GITHUB_PUSH_REPO"],
+    )
+
+
+@cz_blueprint.route("/api/v1/url-submission/test-list/<country_code>", methods=["GET"])
+@role_required(["admin", "user"])
+def get_test_list(country_code):
+    """Fetch citizenlab URL list
+    ---
+    parameters:
+      - in: path
+        name: country_code
+        type: string
+        required: true
+        description: 2-letter country code or "global"
+    responses:
+      200:
+        description: URL list
+        schema:
+          type: object
+          properties:
+            new_entry:
+              type: array
+    """
+    global log
+    log = current_app.logger
+    username = get_username()
+    ulm = get_url_list_manager()
+    tl = ulm.get_test_list(username, country_code)
+    return make_response(jsonify(tl))
+
+
+@cz_blueprint.route("/api/v1/url-submission/add-url", methods=["POST"])
```

Comment:
Perhaps for consistency with the `/api/_/url-submission` endpoint we should do the add-url with the same pattern of having only a `new_entry` an no `old_entry`